src/main/scala/is/hail/annotations/Region.scala

```diff
@@ -3,142 +3,148 @@ package is.hail.annotations
 import java.io.{ObjectInputStream, ObjectOutputStream}
 import java.util
 
-import com.esotericsoftware.kryo.io.{Input, Output}
-import com.esotericsoftware.kryo.{Kryo, KryoSerializable}
 import is.hail.expr.types._
 import is.hail.utils._
 import sun.reflect.generics.reflectiveObjects.NotImplementedException
 
 object Region {
-  def apply(sizeHint: Long = 128): Region = {
-    new Region(new Array[Byte](sizeHint.toInt))
+  def apply(): Region = {
+    val blockSize = 16L * 1024L
+    val blocks = new ArrayBuilder[Long]()
+    blocks += Memory.malloc(blockSize)
+    new Region(
+      blocks,
+      blockSize,
+      0,
+      0L,
+      new ArrayBuilder()
+    )
   }
 
   def scoped[T](f: Region => T): T =
     using(Region())(f)
 }
 
-final class Region(
-  private var mem: Array[Byte],
-  private var end: Long = 0
+final class Region private (
+  private[this] var blocks: ArrayBuilder[Long],
+  private[this] var blockSize: Long,
+  private[this] var activeBlock: Int,
+  private[this] var end: Long,
+  private[this] var bigBlocks: ArrayBuilder[Long]
 ) extends UnKryoSerializable with AutoCloseable {
-  def size: Long = end
+  def summary(): String = {
+    s"""((${blocks.underlying.mkString(" ")})
+        |$blockSize
+        |$end
+        |(${bigBlocks.underlying.mkString(" ")}))""".stripMargin
+  }
 
-  def capacity: Long = mem.length
+  def capacity: Long = blockSize
 
   def copyFrom(other: Region, readStart: Long, writeStart: Long, n: Long) {
-    assert(size <= capacity)
-    assert(other.size <= other.capacity)
+    assert(end <= capacity)
```

Comment:
I think all these checks on (end <= capacity) are going to go pear-shaped as soon
as you start accessing locations in the middle of a big-block allocation.  With
non-contiguous memory and absolute addresses instead of offsets, there isn't much
opportunity for checking the validity of addresses (at least not efficiently - you could
in theory check that the accessed range of addresses lies inside one of the buffers,
but that would involve building some relatively complex data structure and would be too expensive to do e.g. on every loadInt).